File: src/Atlasd/Battlenet/Protocols/Game/ChatCommands/AdminSpoofUserNameCommand.cs
using Atlasd.Battlenet.Protocols.Game.Messages;
using Atlasd.Daemon;
using Atlasd.Localization;
using System;
using System.Collections.Generic;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.ChatCommands
{
    class AdminSpoofUserNameCommand : ChatCommand
    {
        public AdminSpoofUserNameCommand(byte[] rawBuffer, List<string> arguments) : base(rawBuffer, arguments) { }

        public override bool CanInvoke(ChatCommandContext context)
        {
            return context != null && context.GameState != null && context.GameState.ActiveAccount != null;
        }

        public override void Invoke(ChatCommandContext context)
        {
            var r = string.Empty; // reply
            var n1 = Arguments.Count < 1 ? string.Empty : Arguments[0]; // target old name
            var n2 = Arguments.Count < 2 ? string.Empty : Arguments[1]; // target new name

            if (n1.Length == 0 || !Battlenet.Common.GetClientByOnlineName(n1, out var target) || target == null)
            {
                r = Resources.UserNotLoggedOn;
                foreach (var kv in context.Environment)
                    r = r.Replace("{" + kv.Key + "}", kv.Value);
                foreach (var line in r.Split(Battlenet.Common.NewLine))
                    new ChatEvent(ChatEvent.EventIds.EID_ERROR, context.GameState.ChannelFlags, context.GameState.Client.RemoteIPAddress, context.GameState.Ping, context.GameState.OnlineName, line).WriteTo(context.GameState.Client);
                return;
            }

            if (n2.Length == 0 || (n2.Contains("#") && n2[0..n2.IndexOf("#")].Length == 0))
            {
                r = Resources.AdminSpoofUserNameCommandBadValue;
                foreach (var kv in context.Environment)
                    r = r.Replace("{" + kv.Key + "}", kv.Value);
                foreach (var line in r.Split(Battlenet.Common.NewLine))
                    new ChatEvent(ChatEvent.EventIds.EID_ERROR, context.GameState.ChannelFlags, context.GameState.Client.RemoteIPAddress, context.GameState.Ping, context.GameState.OnlineName, line).WriteTo(context.GameState.Client);
                return;
            }

            Arguments.RemoveAt(1); // remove n2
            Arguments.RemoveAt(0); // remove n1
            // Calculates and removes (n1+' '+n2+' ') from (raw) which prints into (newRaw):
            RawBuffer = RawBuffer[(
                Encoding.UTF8.GetByteCount(n1) +
                Encoding.UTF8.GetByteCount(n2) +
                (Arguments.Count > 0 ? 1 : 0) +
                (Arguments.Count > 1 ? 1 : 0))..];

            context.Environment.Add("name1", n1);
            context.Environment.Add("name2", n2);

            // process n1 → n2

            var oldOnlineName = target.OnlineName;
            var oldFlags = target.ChannelFlags;
            var activeChannel = target.ActiveChannel;

            // remove user from their channel
            if (activeChannel != null)
            {
                activeChannel.RemoveUser(target);
            }

            // get a new unique name from n2 (instead of target.Username)
            var searchName = n2.Contains("#") ? n2[0..n2.IndexOf("#")] : n2;
            int serial = 1;

            if (n2.Contains("#"))
            {
                var fields = n2.Split("#");
                int.TryParse(fields[1], out serial);
                if (serial < 1) serial = 1;
            }

            var onlineName = serial == 1 ? searchName : $"{searchName}#{serial}";
            while (!Battlenet.Common.ActiveAccounts.TryAdd(onlineName, target.ActiveAccount)) onlineName = $"{searchName}#{++serial}";
            Battlenet.Common.ActiveAccounts.TryRemove(oldOnlineName, out _);
            target.OnlineName = onlineName;

            // re-key target in active states
            if (!(Battlenet.Common.ActiveGameStates.TryAdd(target.OnlineName, target) &&
                Battlenet.Common.ActiveGameStates.TryRemove(oldOnlineName, out _)))
            {
                Logging.WriteLine(Logging.LogLevel.Error, Logging.LogType.Client, $"Failed to re-key game state from [{oldOnlineName}] to [{target.OnlineName}]");
            }

            // send a new SID_ENTERCHAT to target
            new SID_ENTERCHAT().Invoke(new MessageContext(target.Client, MessageDirection.ServerToClient, new Dictionary<string, object> {}));

            // put them back in their channel
            target.ChannelFlags = oldFlags;
            activeChannel.AcceptUser(target, true);

            // print result to admin
            foreach (var kv in context.Environment)
                r = r.Replace("{" + kv.Key + "}", kv.Value);
            foreach (var line in r.Split(Battlenet.Common.NewLine))
                new ChatEvent(ChatEvent.EventIds.EID_INFO, context.GameState.ChannelFlags, context.GameState.Client.RemoteIPAddress, context.GameState.Ping, context.GameState.OnlineName, line).WriteTo(context.GameState.Client);
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/ChatCommands/AdminSpoofUserNameCommand.vb
Imports AtlasV.Battlenet.Protocols.Game.Messages
Imports AtlasV.Localization
Imports System
Imports System.Collections.Generic
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game.ChatCommands
    Class AdminSpoofUserNameCommand
        Inherits ChatCommand

        Public Sub New(ByVal varRawBuffer As Byte(), ByVal varArguments As List(Of String))
            MyBase.New(varRawBuffer, varArguments)
        End Sub

        Public Overrides Function CanInvoke(ByVal varContext As ChatCommandContext) As Boolean
            Return varContext IsNot Nothing AndAlso varContext.GameState IsNot Nothing AndAlso varContext.GameState.ActiveAccount IsNot Nothing
        End Function

        Public Overrides Sub Invoke(ByVal varContext As ChatCommandContext)
            Dim r As String = String.Empty
            Dim n1 As String = If(Arguments.Count < 1, String.Empty, Arguments(0))
            Dim n2 As String = If(Arguments.Count < 2, String.Empty, Arguments(1))
            Dim target As GameState = Nothing

            If n1.Length = 0 OrElse Not Battlenet.Common.GetClientByOnlineName(n1, target) OrElse target Is Nothing Then
                r = Resources.UserNotLoggedOn

                For Each kv In varContext.Environment
                    r = r.Replace("{" & kv.Key & "}", kv.Value)
                Next

                For Each line In r.Split(Battlenet.Common.NewLine)
                    Call New ChatEvent(ChatEvent.EventIds.EID_ERROR, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, line).WriteTo(varContext.GameState.Client)
                Next

                Return
            End If

            If n2.Length = 0 OrElse (n2.Contains("#") AndAlso n2.Substring(0, n2.IndexOf("#")).Length = 0) Then
                r = Resources.AdminSpoofUserNameCommandBadValue

                For Each kv In varContext.Environment
                    r = r.Replace("{" & kv.Key & "}", kv.Value)
                Next

                For Each line In r.Split(Battlenet.Common.NewLine)
                    Call New ChatEvent(ChatEvent.EventIds.EID_ERROR, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, line).WriteTo(varContext.GameState.Client)
                Next

                Return
            End If

            Arguments.RemoveAt(1)
            Arguments.RemoveAt(0)
            RawBuffer = RawBuffer.Skip((Encoding.UTF8.GetByteCount(n1) + Encoding.UTF8.GetByteCount(n2) + (If(Arguments.Count > 0, 1, 0)) + (If(Arguments.Count > 1, 1, 0)))).ToArray()
            varContext.Environment.Add("name1", n1)
            varContext.Environment.Add("name2", n2)
            Dim oldOnlineName As String = target.OnlineName
            Dim oldFlags As Account.Flags = target.ChannelFlags
            Dim activeChannel As Channel = target.ActiveChannel

            If activeChannel IsNot Nothing Then
                activeChannel.RemoveUser(target)
            End If

            SyncLock Battlenet.Common.ActiveAccounts
                Dim searchName As String = If(n2.Contains("#") = True, n2.Substring(0, n2.IndexOf("#")), n2) '= If(n2.Contains("#") = True, n2(0..n2.IndexOf("#")), n2)
                Dim serial As Integer = 1

                If n2.Contains("#") Then
                    Dim fields = n2.Split("#")
                    Dim unused = Integer.TryParse(fields(1), serial) 'check the return
                    If serial < 1 Then serial = 1
                End If

                Dim onlineName = If(serial = 1, searchName, $"{searchName}#{serial}")

                While Battlenet.Common.ActiveAccounts.ContainsKey(onlineName.ToLower())
                    onlineName = $"{searchName}#{System.Threading.Interlocked.Increment(serial)}"
                End While

                SyncLock target.ActiveAccount
                    Battlenet.Common.ActiveAccounts.Remove(oldOnlineName.ToLower())
                    target.OnlineName = onlineName
                    Battlenet.Common.ActiveAccounts.Add(onlineName.ToLower(), target.ActiveAccount)
                End SyncLock
            End SyncLock

            SyncLock Battlenet.Common.ActiveGameStates
                Battlenet.Common.ActiveGameStates.Remove(oldOnlineName.ToLower())
                Battlenet.Common.ActiveGameStates.Add(target.OnlineName, target)
            End SyncLock

            '// send a New SID_ENTERCHAT to target
            Call New SID_ENTERCHAT().Invoke(New MessageContext(target.Client, MessageDirection.ServerToClient, New Dictionary(Of String, Object)()))

            target.ChannelFlags = oldFlags
            activeChannel.AcceptUser(target, True)

            For Each kv In varContext.Environment
                r = r.Replace("{" & kv.Key & "}", kv.Value)
            Next

            For Each line In r.Split(Battlenet.Common.NewLine)
                Call New ChatEvent(ChatEvent.EventIds.EID_INFO, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, line).WriteTo(varContext.GameState.Client)
            Next
        End Sub
    End Class
End Namespace
